Products/ZenCollector/daemon.py

```diff
@@ -832,53 +832,51 @@ def _startMaintenance(self, ignored=None):
         maintenanceCycle = MaintenanceCycle(interval, heartbeatSender, self._maintenanceCycle)
         maintenanceCycle.start()
 
+    @defer.inlineCallbacks
     def _maintenanceCycle(self, ignored=None):
         """
         Perform daemon maintenance processing on a periodic schedule. Initially
-        called after the daemon configuration loader task is added, but afterward
-        will self-schedule each run.
+        called after the daemon configuration loader task is added,
+        but afterward will self-schedule each run.
         """
         self.log.debug("Performing periodic maintenance")
+        ret = None
+        try:
+            if not self.options.cycle:
+                ret = "No maintenance required"
+            elif getattr(self.preferences, 'pauseUnreachableDevices', True):
+                # TODO: handle different types of device issues, such as WMI issues
+                issues = yield self.getDevicePingIssues()
+                ret = yield self._processDeviceIssues(issues)
+        except Exception as err:
+            self.log.exception('failure in _maintenanceCycle')
+            raise err
 
-        def _processDeviceIssues(result):
-            self.log.debug("deviceIssues=%r", result)
-            if result is None:
-                return result  # exception or some other problem
-
-            # Device ping issues returns as a tuple of (deviceId, count, total)
-            # and we just want the device id
-            newUnresponsiveDevices = set(i[0] for i in result)
-
-            clearedDevices = self._unresponsiveDevices.difference(newUnresponsiveDevices)
-            for devId in clearedDevices:
-                self.log.debug("Resuming tasks for device %s", devId)
-                self._scheduler.resumeTasksForConfig(devId)
-
-            self._unresponsiveDevices = newUnresponsiveDevices
-            for devId in self._unresponsiveDevices:
-                self.log.debug("Pausing tasks for device %s", devId)
-                self._scheduler.pauseTasksForConfig(devId)
-
-            return result
-
-        def _getDeviceIssues():
-            # TODO: handle different types of device issues, such as WMI issues
-            d = self.getDevicePingIssues()
-            return d
-
-        def _maintenance():
-            if self.options.cycle:
-                if getattr(self.preferences, 'pauseUnreachableDevices', True):
-                    d = defer.maybeDeferred(_getDeviceIssues)
-                    d.addCallback(_processDeviceIssues)
-                else:
-                    d = defer.succeed(None)
-            else:
-                d = defer.succeed("No maintenance required")
-            return d
+        defer.returnValue(ret)
 
-        d = _maintenance()
-        return d
+    @defer.inlineCallbacks
+    def _processDeviceIssues(self, result):
```

Comment:
There's no I/O occurring in this method, so there's no need for it to use inlineCallbacks.

Reply:
Fixed